### DISCUSSION TITLE
LinkerD with Headless Service or not?

### DISCUSSION THREAD

**Discussion by pantinor** [linkerd/linkerd2#5985] [Q&A]
In reference to this discussion about the request level load balancing across dynamically scheduled servers and pods:

https://kubernetes.io/blog/2018/11/07/grpc-load-balancing-on-kubernetes-without-tears/

Is it recommended to **not use a headless service?** and instead use LinkerD and lightweight proxy to load balance the requests cross dynamic pods and gRPC servers in the cluster?

**Comment by wmorgan** [comment]
Linkerd can only provide load-balancing to regular (non-headless) services. If you have a gRPC service, my recommendation is to make it regular service and use Linkerd. :)

**Reply by pantinor** [reply]
OK, thats fine.  But in this case, for the gRPC client to send a request, the channel shall be established to which host?  Can the client target any of the addresses?  Normally with a cluster service we would target the clusterIP and port to load balance.  SO what shall be the target host when balancing with LinkerD, or does it not matter ?

**Reply by olix0r** [reply]
When an application connects to a service's `clusterIP`, the linkerd proxy will do load balancing over the pods in the service (and discover service profiles / traffic splits for the service). When a client targets a pod, linkerd will just forward requests to the pod without doing any service-level routing/loadbalancing/etc.

To be clear: this has nothing to do with the `:authority` header value and relies solely on the IP address to which the client builds a TCP connection.
